### PR TITLE
[WIP] Revamping of timeline to use different and more efficient api

### LINKED ISSUE #452: [FR] add editing for gists
**App Version: 2.1.0**

**OS Version: 25**

**Model: HTC-Nexus 9**

 One day when you get around to it (maybe on a holiday or something) you add the ability to edit gists

### LINKED ISSUE #781: Scroll bar placement
**FastHub Version: 4.0.3**  
**Android Version: 7.1.2 (SDK: 25)**  
**Device Information:**  
- MANUFACTURER: Google
- BRAND: google  
- MODEL: Pixel  
---

 When scrolling through an issue, the scroll bar doesn't always start at the top and end at the bottom. 

Top

Bottom

### LINKED ISSUE #809: Horizontal rulers are not displayed
**FastHub Version: 4.0.3**  
**Android Version: 7.1.2 (SDK: 25)**  
**Device Information:**  
- MANUFACTURER: samsung
- BRAND: samsung  
- MODEL: SM-G930F  
---

Horizontal rulers are not displayed. When looking at issues and also in the previews, I nowhere see the horizontal rulers that definitely are there and also are shown in the browser.

### LINKED ISSUE #814: Display build statuses for commits.
Just another feature request.  
https://developer.github.com/v3/repos/statuses/

### LINKED ISSUE #825: only public entries on an organisations feed
**FastHub Version:**  4.0.3
**Android Version:**  7.1.2
**Device Information:**  
- MANUFACTURER: OnePlus  
- BRAND: OnePlus  
- MODEL: One  

When i open the feed of an organisation, only entries that belong to public repositories of that organisation show up. Entries belonging to private repositories of that organisation are not visible.

Is that intended behaviour, or a bug?

### PR DESCRIPTION
**Pull request by yakov116**
closes #452 closes #781 fixes #809 closes #814 fixes #825
#### Cross reference 
   - [x] Add cross referenced to issues and pr's
        - [x] Fix clicking link on cross references 
        - [ ] Add title to the cross reference ` "title": "Emulator detection",` (see number #712)
        - [ ] fix cross reference lang when it has a commit `"event": "referenced",
        "commit_id": "1b922e9e17963ccdbde21e1255242d984fadf031"`
        - [ ] Add pull request lang for cross reference
        - [x] Add repo if the reference was from an external repo 
#### Pr Issues
   - [x] Add back PR statuses
   - [x] Fix Close and reopen (always says merged)
   - [ ] Fix Merge being double (don't show referenced)
   - [x] Fix reply and share
##### Side issues
   - [x] After closing an issue refresh
   - [x] Fix Pictures being very large (issue after lib change)
   - [ ] Fix readme pull down refresh

### CODE REVIEW

File: app/src/main/java/com/fastaccess/data/dao/IssuesPageable.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    public int next;
+    public int prev;
+    public int last;
+    public int totalCount;
+    public boolean incompleteResults;
+    public List<M> items;
+
+    @Override public int describeContents() { return 0; }
+
+    @Override public void writeToParcel(Parcel dest, int flags) {
```
Comment (k0shk0sh) [inline app/src/main/java/com/fastaccess/data/dao/IssuesPageable.java]: Test

Comment (k0shk0sh) [inline app/src/main/java/com/fastaccess/data/dao/IssuesPageable.java]: test 2

File: app/src/main/java/com/fastaccess/data/dao/IssuesPageable.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+/**
+ * Created by Kosh on 15 Nov 2016, 7:04 PM
+ */
+
+
+@Getter @Setter @NoArgsConstructor
+public class IssuesPageable<M> implements Parcelable {
```
Comment (k0shk0sh) [inline app/src/main/java/com/fastaccess/data/dao/IssuesPageable.java]: Test

Comment (k0shk0sh) [inline app/src/main/java/com/fastaccess/data/dao/IssuesPageable.java]: Hello World

File: app/src/main/java/com/fastaccess/data/dao/timeline/AuthorModel.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    @Override public int describeContents() { return 0; }
+
+    @Override public void writeToParcel(Parcel dest, int flags) {
+        dest.writeString(this.name);
+        dest.writeString(this.email);
+        dest.writeLong(this.date != null ? this.date.getTime() : -1);
+    }
+
+    protected AuthorModel(Parcel in) {
+        this.name = in.readString();
```
Comment (yakov116) [inline app/src/main/java/com/fastaccess/data/dao/timeline/AuthorModel.java]: Testing single line comment

**Comment by yakov116** [comment]
Test

File: app/src/main/java/com/fastaccess/data/dao/timeline/PullRequestTimelineModel.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    public PullRequestTimelineModel(PullRequestTimelineQuery.Node node) {
+        this.node = node;
+    }
+
+    public int getType() {
+        if (pullRequest != null) return HEADER;
+        if (node.asAssignedEvent() != null || node.asClosedEvent() != null
+                || node.asDemilestonedEvent() != null || node.asHeadRefDeletedEvent() != null
```
Comment (yakov116) [inline app/src/main/java/com/fastaccess/data/dao/timeline/PullRequestTimelineModel.java]: where is `node.asHeadRefRestoredEvent()` ?

File: app/src/main/java/com/fastaccess/ui/adapter/viewholder/PullRequestEventViewHolder.kt [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+            avatarLayout.setUrl(it.avatarUrl().toString(), it.login(), false, LinkParserHelper.isEnterprise(it.url().toString()))
+        }
+    }
+
+    private fun renamedEvent(event: PullRequestTimelineQuery.AsRenamedTitleEvent) {
+        event.actor()?.let {
+            stateText.text = SpannableBuilder.builder()
+                    .bold(it.login())
+                    .append(" ")
+                    .append("renamed this from").append(" ").append(event.previousTitle())
```
Comment (yakov116) [inline app/src/main/java/com/fastaccess/ui/adapter/viewholder/PullRequestEventViewHolder.kt]: changed the title from

File: app/src/main/java/com/fastaccess/ui/adapter/viewholder/PullRequestEventViewHolder.kt [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+            avatarLayout.setUrl(it.avatarUrl().toString(), it.login(), false, LinkParserHelper.isEnterprise(it.url().toString()))
+        }
+    }
+
+    private fun referenceEvent(event: PullRequestTimelineQuery.AsReferencedEvent) {
+        event.actor()?.let {
+            stateText.text = SpannableBuilder.builder()
+                    .bold(it.login())
+                    .append(" ")
+                    .append(if (event.isCrossRepository) "cross referenced this" else "referenced this")
```
Comment (yakov116) [inline app/src/main/java/com/fastaccess/ui/adapter/viewholder/PullRequestEventViewHolder.kt]: This is going to need work.

We will need to know if this is a pull request or not. Leave it for me,

File: app/src/main/java/com/fastaccess/ui/adapter/viewholder/PullRequestEventViewHolder.kt [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+            avatarLayout.setUrl(it.avatarUrl().toString(), it.login(), false, LinkParserHelper.isEnterprise(it.url().toString()))
+        }
+    }
+
+    private fun mergedEvent(event: PullRequestTimelineQuery.AsMergedEvent) {
+        event.actor()?.let {
+            stateText.text = SpannableBuilder.builder()
+                    .bold(it.login())
+                    .append(" ")
+                    .append("merged this in")
```
Comment (yakov116) [inline app/src/main/java/com/fastaccess/ui/adapter/viewholder/PullRequestEventViewHolder.kt]: merged commit `pull request commit` into `headref`

File: app/src/main/java/com/fastaccess/ui/adapter/viewholder/PullRequestEventViewHolder.kt [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+            avatarLayout.setUrl(it.avatarUrl().toString(), it.login(), false, LinkParserHelper.isEnterprise(it.url().toString()))
+        }
+    }
+
+    private fun headRefRestoredEvent(event: PullRequestTimelineQuery.AsHeadRefRestoredEvent) {
+        event.actor()?.let {
+            stateText.text = SpannableBuilder.builder()
+                    .bold(it.login())
+                    .append(" ")
+                    .append("head ref restored", BackgroundColorSpan(HtmlHelper.getWindowBackground(PrefGetter.getThemeType())))
```
Comment (yakov116) [inline app/src/main/java/com/fastaccess/ui/adapter/viewholder/PullRequestEventViewHolder.kt]: you didn't commit the change?

File: app/src/main/java/com/fastaccess/ui/adapter/viewholder/PullRequestEventViewHolder.kt [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+            avatarLayout.setUrl(it.avatarUrl().toString(), it.login(), false, LinkParserHelper.isEnterprise(it.url().toString()))
+        }
+    }
+
+    private fun refDeletedEvent(event: PullRequestTimelineQuery.AsHeadRefDeletedEvent) {
+        event.actor()?.let {
+            stateText.text = SpannableBuilder.builder()
+                    .bold(it.login())
+                    .append(" ")
+                    .append("head ref delete", BackgroundColorSpan(HtmlHelper.getWindowBackground(PrefGetter.getThemeType())))
```
Comment (yakov116) [inline app/src/main/java/com/fastaccess/ui/adapter/viewholder/PullRequestEventViewHolder.kt]: Same here the string comes out

File: app/src/main/java/com/fastaccess/ui/adapter/viewholder/PullRequestEventViewHolder.kt [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+            avatarLayout.setUrl(it.avatarUrl().toString(), it.login(), false, LinkParserHelper.isEnterprise(it.url().toString()))
+        }
+    }
+
+    private fun deployedEvent(event: PullRequestTimelineQuery.AsDeployedEvent) {
+        event.actor()?.let {
+            stateText.text = SpannableBuilder.builder()
+                    .bold(it.login())
+                    .append(" ")
+                    .append("make a deployment", BackgroundColorSpan(HtmlHelper.getWindowBackground(PrefGetter.getThemeType())))
```
Comment (yakov116) [inline app/src/main/java/com/fastaccess/ui/adapter/viewholder/PullRequestEventViewHolder.kt]: **made** a deployment

**Comment by yakov116** [comment]
The commit should say login not name 

_Sent from my Google Nexus 9 using [FastHub Debug](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

File: app/src/main/java/com/fastaccess/data/dao/timeline/AuthorModel.java (hunk 3 above)
Comment (k0shk0sh) [inline app/src/main/java/com/fastaccess/data/dao/timeline/AuthorModel.java]: Testing reply

File: app/src/main/java/com/fastaccess/data/dao/timeline/PullRequestTimelineModel.java (hunk 4 above)
Comment (yakov116) [inline app/src/main/java/com/fastaccess/data/dao/timeline/PullRequestTimelineModel.java]: test

File: app/src/main/java/com/fastaccess/data/dao/CommentRequestModel.java [hunk 11]
```diff
@@ -25,7 +27,7 @@ public CommentRequestModel() {}
         if (this == o) return true;
         if (o == null || getClass() != o.getClass()) return false;
         CommentRequestModel that = (CommentRequestModel) o;
-        return position == that.position && (path != null ? path.equals(that.path) : that.path == null);
+        return Objects.equals(position, that.position) && (path != null ? path.equals(that.path) : that.path == null);
```
Comment (k0shk0sh) [inline app/src/main/java/com/fastaccess/data/dao/CommentRequestModel.java]: YANKI OBJECTS is JAVA 8 FEATURE & THAT MEANS IT WONT WORK ON ANY ANDROID BUT LATEST!!!!!!!!!

**Pull request closed without merging**